x/distribution/abci_test.go

```diff
@@ -0,0 +1,107 @@
+
+import (
+	"testing"
+
+	"github.com/cosmos/cosmos-sdk/crypto/keys/ed25519"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	"github.com/cosmos/cosmos-sdk/simapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/staking/teststaking"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	"github.com/stretchr/testify/require"
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+)
+
+const (
+	totalValidators  = 6
+	lazyValidatorIdx = 2
+	power            = 100 / totalValidators
+)
+
+var (
+	valTokens                = sdk.TokensFromConsensusPower(50, sdk.DefaultPowerReduction)
+	validatorCommissionRates = stakingtypes.NewCommissionRates(sdk.OneDec(), sdk.OneDec(), sdk.OneDec())
+)
+
+type validator struct {
+	addr   sdk.ValAddress
+	pubkey cryptotypes.PubKey
+	votes  []abci.VoteInfo
+}
+
+// Context in https://github.com/cosmos/cosmos-sdk/issues/9161
+func TestVerifyProposerRewardAssignement(t *testing.T) {
+	app := simapp.Setup(t, false)
+	ctx := app.BaseApp.NewContext(false, tmproto.Header{})
+	addrs := simapp.AddTestAddrsIncremental(app, ctx, totalValidators, valTokens)
+	tstaking := teststaking.NewHelper(t, ctx, app.StakingKeeper)
+	tstaking.Commission = validatorCommissionRates
+
+	// create validators
+	validators := make([]validator, totalValidators-1)
+	for i := range validators {
+		validators[i].addr = sdk.ValAddress(addrs[i])
+		validators[i].pubkey = ed25519.GenPrivKey().PubKey()
+		validators[i].votes = make([]abci.VoteInfo, totalValidators)
+		tstaking.CreateValidatorWithValPower(validators[i].addr, validators[i].pubkey, power, true)
+	}
+	app.EndBlock(abci.RequestEndBlock{})
+	require.NotEmpty(t, app.Commit())
+
+	// verify validators lists
+	require.Len(t, app.StakingKeeper.GetAllValidators(ctx), totalValidators)
+	for i, val := range validators {
+		// verify all validator exists
+		require.NotNil(t, app.StakingKeeper.ValidatorByConsAddr(ctx, sdk.GetConsAddress(val.pubkey)))
+
+		// populate last commit info
+		voteInfos := []abci.VoteInfo{}
+		for _, val2 := range validators {
+			voteInfos = append(voteInfos, abci.VoteInfo{
+				Validator: abci.Validator{
+					Address: sdk.GetConsAddress(val2.pubkey),
+					Power:   power,
+				},
+				SignedLastBlock: true,
+			})
+		}
+
+		// have this validator only submit the minimum amount of pre-commits
+		if i == lazyValidatorIdx {
+			for j := totalValidators * 2 / 3; j < len(voteInfos); j++ {
+				voteInfos[j].SignedLastBlock = false
+			}
+		}
+
+		validators[i].votes = voteInfos
+	}
+
+	// previous block submitted by validator n-1 (with 100% previous commits) and proposed by lazy validator
+	app.BeginBlock(abci.RequestBeginBlock{
+		Header:         tmproto.Header{Height: app.LastBlockHeight() + 1, ProposerAddress: sdk.GetConsAddress(validators[lazyValidatorIdx].pubkey)},
+		LastCommitInfo: abci.LastCommitInfo{Votes: validators[lazyValidatorIdx-1].votes},
```

Comment:
if this is the block proposed by the lazy validator, shouldnt it not have 100% precommits? seems like an off by one error that masks the off by one error i claim is in the code

Reply:
when a validator proposes a block, they include the precommits they've seen from the last block. they are supposed to wait before proposing to see more precommits. so when they propose block N, it includes precommits from N-1, and when the block N is committed, the BeginBlock.LastCommitInfo should contain the signatures for block N-1 included by the proposer of block N. In this test, the lazyValidator seems to be including 100% of the signatures. but looking at the reward they get, its small compared to the others, which suggests she's only being rewards for the signatures included in the next block, which is exactly the bug I was trying to highlight!